src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+struct TagBacktrace {
+    backtrace: Vec<CurrentTag>,
+}
+
+impl TagBacktrace {
+    fn push_back(&mut self, tag: CurrentTag) {
+        self.backtrace.push(tag);
+    }
+
+    /// Remove the top tag and return it. Return `Err(())` on tag mismatching.
+    fn collide(&mut self, name: &[u8]) -> Result<CurrentTag, ()> {
+        if self
+            .backtrace
+            .last()
+            .map(|t| t.matches(name))
+            .unwrap_or_default()
+        {
+            // There is at least one element in backtrace, hence this unwrap is safe.
+            Ok(self.backtrace.pop().unwrap())
+        } else {
+            Err(())
+        }
+    }
+
+    fn top_mut(&mut self) -> Option<&mut CurrentTag> {
+        self.backtrace.last_mut()
+    }
+}
+
+/// The tag we are scanning, with additional states.
+enum CurrentTag {
+    TraceQueryResult(TraceQueryResultState),
+    Node(NodeState),
+    Row(RowState),
+    Backtrace(BacktraceState),
+    Frame(FrameState),
+    Other(String),
+}
+
+impl CurrentTag {
+    fn matches(&self, name: &[u8]) -> bool {
+        match name {
+            TRACE_QUERY_RESULT => matches!(self, Self::TraceQueryResult(_)),
+            NODE => matches!(self, Self::Node(_)),
+            ROW => matches!(self, Self::Row(_)),
+            BACKTRACE => matches!(self, Self::Backtrace(_)),
+            FRAME => matches!(self, Self::Frame(_)),
+            other => matches!(self, Self::Other(tag) if tag.as_bytes() == other),
+        }
+    }
+}
+
+#[derive(Default)]
+struct TraceQueryResultState {
+    nodes: Vec<Node>,
+}
+
+#[derive(Default)]
+struct NodeState {
+    rows: Vec<Row>,
+}
+
+#[derive(Default)]
+struct RowState {
+    backtrace: Option<Arc<Backtrace>>,
+}
+
+#[derive(Default)]
+struct BacktraceState {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+#[derive(Default)]
+struct FrameState {
+    id: u64,
+    name: Vec<u8>,
+}
+
+struct Node {
+    rows: Vec<Row>,
+}
+
+struct Row {
+    backtrace: Arc<Backtrace>,
+}
+
+struct Backtrace {
+    id: u64,
+    frames: Vec<Arc<Frame>>,
+}
+
+struct Frame {
+    id: u64,
+    name: Vec<u8>,
+}
+
+impl Backtrace {
+    fn to_folded(&self) -> Vec<u8> {
+        let mut folded = Vec::new();
+        let mut first = Some(());
+        // Because stack frames are arranged from top to bottom in xctrace's
+        // output, here we use `.rev(`.
+        for frame in self.frames.iter().rev() {
+            if first.take().is_none() {
+                folded.push(b';');
+            }
+            let frame_name = String::from_utf8_lossy(&frame.name);
+            let frame_name = fix_partially_demangled_rust_symbol(&frame_name);
+            folded.extend(frame_name.as_bytes().iter().copied());
+        }
+        folded
+    }
+}
+
+/// Unescapes the text in xml exported from xctrace.
+fn unescape_xctrace_text(x: Cow<'_, [u8]>) -> Vec<u8> {
+    // xctrace shouldn't give us invalid xml text here, therefore
+    // we don't expect the error branch being hit:
```

Comment:
If we don't expect to hit the error case, should we have it actually error instead?